### PR TITLE
PXP-628: [CLI]: Display new release info banner

### PR DESCRIPTION
**Pull request by mfauzaan**
## Summary
To let developer know that there’s a new Wukong CLI version, we should display the notification in the terminal whenever user trigger the CLI. 

To avoid spam, we should only display this maximum every 24hours 

Ticket: https://mindvalley.atlassian.net/browse/PXP-624

## What's Changed
Screenshot 2023-11-01 at 4 26 47 PM

### CODE REVIEW

**Comment by sauron-droid** [comment]
[APPROVALNOTIFIER] This PR is **NOT APPROVED**

This pull-request has been approved by: *<a href="https://github.com/mindvalley/wukong-cli/pull/164#" title="Author self-approved">mfauzaan</a>*


The full list of commands accepted by this bot can be found [here](https://go.k8s.io/bot-commands?repo=mindvalley%2Fwukong-cli).

<details open>
Needs approval from an approver in each of these files:

- **[OWNERS](https://github.com/mindvalley/wukong-cli/blob/main/OWNERS)**

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>

File: cli/src/update.rs
Comment (jk-gan) [inline cli/src/update.rs:69]: Not really sure if I like this or not, if we are not returning any error in this function, then I guess we don't need `Result` type, just `Option<ReleaseInfo>` will do.

File: cli/src/update.rs
Comment (jk-gan) [inline cli/src/update.rs:111]: why the `github_api_url` param need to be an `Option`?

File: cli/src/update.rs
Comment (jk-gan) [inline cli/src/update.rs:133]: the error message should always be using `eprintln!` to printout on the `stderr` channel.

File: cli/src/config.rs
Comment (jk-gan) [inline cli/src/config.rs:104]: What is the purpose of storing all these informations into the config file? from what I understand, we only need the `checked_for_update_at` value.

Comment (jk-gan) [reply cli/src/config.rs:104]: Btw I think `checked_for_update_at` is not a good name. I suggest change it to something like `last_checked_at` or `last_update_checked_at`.

File: cli/src/config.rs
Comment (jk-gan) [inline cli/src/config.rs:58]: I think we can call it `update_check` or `new_update_check`, so in the config file it will be:
```toml
[update_check]
last_checked_at = "xxx"
```

Comment (jk-gan) [reply cli/src/config.rs:58]: what do you think?

**Review by jk-gan — CHANGES_REQUESTED** [review]

**Comment by jk-gan** [comment]
Quick question, do we want to show error if there is an error occur during the update check? What is the UX you want to bring to the users?

File: cli/src/update.rs
Comment (jk-gan) [inline cli/src/update.rs:35]: I believe the ticket and your PR description mentions `To avoid spam, we should only display this maximum 1 times per week.`, so this line is wrong

**Review by jk-gan — CHANGES_REQUESTED** [review]

File: cli/src/update.rs
Comment (mfauzaan) [reply cli/src/update.rs:111]: aah, this is because i wanna make sure this function is testable. I need to pass the baseURL to this function

File: cli/src/config.rs
Comment (mfauzaan) [reply cli/src/config.rs:58]: im thinking to keep as release_info, reason being is that im storing the release info in here not only update_check. The intend for now is to check for updates for now but we might have a future use case and keeping the naming as release info makes more sense. 

And i've updated the logic; the idea is regardless of github response we take the result from release info and build the logic so if there is a failure on github we still compare with the latest one on config because version always goes up

File: cli/src/update.rs
Comment (jk-gan) [reply cli/src/update.rs]: but this still not explain why the param should be `Option`, it can be `&str` and still testable

Comment (mfauzaan) [reply cli/src/update.rs]: Here is the code, 
` github_api_url.unwrap_or(GITHUB_API_URL)` not sure if we can use unwrap_or if we do not use Option?

File: cli/src/config.rs
Comment (jk-gan) [reply cli/src/config.rs:58]: I still don't see why we need to store the values here, if there is a failure on GitHub api, then we shouldn't show the *potentially outdated* info to the user, we should just show nothing. 

What are the future use case that need these values?

**Review by jk-gan — CHANGES_REQUESTED** [review]

File: cli/src/config.rs
Comment (mfauzaan) [reply cli/src/config.rs:58]: So my understanding of version number is always are incremental so last cached value is higher then the current version i wouldn't say its really outdated data just that there is a chance that there might be a higher version.

But doesnt mean the upgrade message shows isn't true. So circling back to main intention of showing the upgrade option is that user is using an outdated version

Comment (mfauzaan) [reply cli/src/config.rs:58]: And future uses cases i can think of now is, giving options to the user to upgrade within out CLI
As we have all the data we need its just matter of handling the update

File: cli/src/update.rs
Comment (jk-gan) [reply cli/src/update.rs]: yeah, the thing is you are not even using the `None` case, hence the question "why this is an `Option`?"

**Review by jk-gan — APPROVED** [review]

**Comment by sauron-droid** [comment]
[APPROVALNOTIFIER] This PR is **NOT APPROVED**

This pull-request has been approved by: *<a href="https://github.com/mindvalley/wukong-cli/pull/164#pullrequestreview-1714193099" title="Approved">jk-gan</a>*, *<a href="https://github.com/mindvalley/wukong-cli/pull/164#" title="Author self-approved">mfauzaan</a>*


The full list of commands accepted by this bot can be found [here](https://go.k8s.io/bot-commands?repo=mindvalley%2Fwukong-cli).

<details open>
Needs approval from an approver in each of these files:

- **[OWNERS](https://github.com/mindvalley/wukong-cli/blob/main/OWNERS)**

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>

**Comment by sauron-droid** [comment]
[APPROVALNOTIFIER] This PR is **NOT APPROVED**

This pull-request has been approved by: *<a href="https://github.com/mindvalley/wukong-cli/pull/164#" title="Author self-approved">mfauzaan</a>*


The full list of commands accepted by this bot can be found [here](https://go.k8s.io/bot-commands?repo=mindvalley%2Fwukong-cli).

<details open>
Needs approval from an approver in each of these files:

- **[OWNERS](https://github.com/mindvalley/wukong-cli/blob/main/OWNERS)**

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>

**Comment by sauron-droid** [comment]
[APPROVALNOTIFIER] This PR is **NOT APPROVED**

This pull-request has been approved by: *<a href="https://github.com/mindvalley/wukong-cli/pull/164#" title="Author self-approved">mfauzaan</a>*


The full list of commands accepted by this bot can be found [here](https://go.k8s.io/bot-commands?repo=mindvalley%2Fwukong-cli).

<details open>
Needs approval from an approver in each of these files:

- **[OWNERS](https://github.com/mindvalley/wukong-cli/blob/main/OWNERS)**

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>

File: cli/src/update.rs
Comment (onimsha) [inline cli/src/update.rs:61]:
```suggestion
        eprintln!("To upgrade, run: brew update && brew upgrade wukong");
```

**Review by onimsha — DISMISSED** [review]

**Comment by sauron-droid** [comment]
[APPROVALNOTIFIER] This PR is **NOT APPROVED**

This pull-request has been approved by: *<a href="https://github.com/mindvalley/wukong-cli/pull/164#" title="Author self-approved">mfauzaan</a>*, *<a href="https://github.com/mindvalley/wukong-cli/pull/164#pullrequestreview-1714428140" title="Approved">onimsha</a>*


The full list of commands accepted by this bot can be found [here](https://go.k8s.io/bot-commands?repo=mindvalley%2Fwukong-cli).

<details open>
Needs approval from an approver in each of these files:

- **[OWNERS](https://github.com/mindvalley/wukong-cli/blob/main/OWNERS)**

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>

**Comment by sauron-droid** [comment]
[APPROVALNOTIFIER] This PR is **NOT APPROVED**

This pull-request has been approved by: *<a href="https://github.com/mindvalley/wukong-cli/pull/164#" title="Author self-approved">mfauzaan</a>*


The full list of commands accepted by this bot can be found [here](https://go.k8s.io/bot-commands?repo=mindvalley%2Fwukong-cli).

<details open>
Needs approval from an approver in each of these files:

- **[OWNERS](https://github.com/mindvalley/wukong-cli/blob/main/OWNERS)**

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>

**Comment by sauron-droid** [comment]
[APPROVALNOTIFIER] This PR is **NOT APPROVED**

This pull-request has been approved by: *<a href="https://github.com/mindvalley/wukong-cli/pull/164#" title="Author self-approved">mfauzaan</a>*


The full list of commands accepted by this bot can be found [here](https://go.k8s.io/bot-commands?repo=mindvalley%2Fwukong-cli).

<details open>
Needs approval from an approver in each of these files:

- **[OWNERS](https://github.com/mindvalley/wukong-cli/blob/main/OWNERS)**

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>

**Comment by sauron-droid** [comment]
[APPROVALNOTIFIER] This PR is **NOT APPROVED**

This pull-request has been approved by: *<a href="https://github.com/mindvalley/wukong-cli/pull/164#" title="Author self-approved">mfauzaan</a>*, *<a href="https://github.com/mindvalley/wukong-cli/pull/164#pullrequestreview-1714460626" title="Approved">onimsha</a>*


The full list of commands accepted by this bot can be found [here](https://go.k8s.io/bot-commands?repo=mindvalley%2Fwukong-cli).

<details open>
Needs approval from an approver in each of these files:

- **[OWNERS](https://github.com/mindvalley/wukong-cli/blob/main/OWNERS)**

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>

**Pull request merged**
